[FIX] account_journal_general_sequence fix multicompany

anddago78 opened the pull request:
This module extends the functionality of account to support sequenced entry
numbers and to allow you to renumber them if needed, on demand, 
this fix make wich multicompany work correctly.

MT-3076 @moduon @rafaelbn @Shide @yajo @fcvalgar  please reviews :)

fcvalgar reviewed (approved):
Functional review.

LGTM thank you @anddago78

rafaelbn reviewed (approved):
Tested in local 👍🏼

Shide commented on account_journal_general_sequence/models/account_journal.py:
```diff
@@ -13,27 +13,20 @@ class AccountJournal(models.Model):
         string="Account entry number sequence",
-        default=lambda self: self._default_entry_number_sequence(),
+        compute="_compute_entry_number_sequence",
         copy=False,
```
Maybe add `store=True`, `readonly=False` to allow select other sequence? Also add domain for free chosing sequences on a secure multi-company environment.
```suggestion
        copy=False,
        store=True,
        readonly=False,
        domain="[('company_id', '=', company_id)]",
```

Shide commented on account_journal_general_sequence/models/account_journal.py:
```diff
@@ -13,27 +13,20 @@ class AccountJournal(models.Model):
-        return result
+    def _compute_entry_number_sequence(self):
+        """Compute sequence for all journals."""
+        for rec in self:
+            sequence = False
+            if rec.company_id.name:
+                sequence = self.env["ir.sequence"].search(
+                    [
+                        ("code", "=", "account_journal_general_sequence.default"),
+                        ("company_id", "=", rec.company_id.id),
+                    ]
+                )
+            rec.entry_number_sequence_id = sequence.id
```
Check if the chosen sequence is on another company instead of checking the name of the company and remove the default false value because we are going to do something if it's incorrect or it's not set.

Also, when assigning the sequence, can throw an exception if company name is falsy (can't happen for other constraint). Instead, assign the sequence only if the "if" happens (one more tab).

```suggestion
        is_model = self.env["ir.sequence"]
        for rec in self:
            if rec.entry_number_sequence_id.company_id != rec.company_id:
                rec.entry_number_sequence_id = is_model.search(
                    [
                        ("code", "=", "account_journal_general_sequence.default"),
                        ("company_id", "=", rec.company_id.id),
                    ], limit=1
                ).id
```

yajo replied:
According to my other comments, this will not be necessary because we shouldn't be using a compute after all.

Shide reviewed (commented):
Great job, this request of changes are if we don't want to be too strict in chosing sequences, but maybe it's not aligned with the way we want to work. @rafaelbn @yajo please correct me if i'm wrong

yajo commented on account_journal_general_sequence/models/account_journal.py:
```diff
@@ -13,27 +13,20 @@ class AccountJournal(models.Model):
         comodel_name="ir.sequence",
         string="Account entry number sequence",
-        default=lambda self: self._default_entry_number_sequence(),
+        compute="_compute_entry_number_sequence",
```
IMHO you shouldn't have changed this default for a compute. The previous implementation was better because:

- It was just adding a default. If a user wanted to have a different sequence for another journal, they just had to change it. We'd be now missing this feature.
- It was making sure a sequence exists for the company when actually being used. See my comment on the `pre_init_companies` method for more details.

yajo commented on account_journal_general_sequence/hooks.py:
```diff
@@ -0,0 +1,22 @@
+def pre_init_companies(cr):
+    env = api.Environment(cr, SUPERUSER_ID, {})
+    cmps = env["res.company"].search([])
+    for company in cmps:
```
This is only going to be executed at module install time. It will not run on update. Imagine this story:

1. Database contains company 1 and company 2.
2. Admin installs this module. It creates sequences for companies 1 and 2.
3. Admin creates company 3.

With this implementation, Company 3 will never have a sequence. That's why the previous implementation was more sound.

yajo reviewed (changes requested):
Hello! There's something I don't understand here.

In the previous implementation, the sequence was created on demand whenever used for the 1st time. However, with this new implementation, a sequence is created per company at module install (*not* at updates) and at company creation.

What I don't understand is... where's the benefit? 😅  All that was needed was that the sequence was by company. Wasn't enough just by adding the `"company_id": self.env.company.id` to the sequence creation without having to change the whole implementation?

With this new implementation, if you don't want to use the sequence that is created by default, because you want a different sequence in some journal, you now can't do that. So we lost a feature in the way.

Also, why do tests have to change? This one doesn't seem like a backwards-incompatible change.

Finally, it's lacking a migration script as you already commented.

Please read carefully the task ADP. We can clarify it later by meet if you want. Please set this as draft in the mean time.

Thanks!

rafaelbn commented:
@yajo the previous implementation is incorrect, was a bug. It should've like that from the very beginning

It was a missunderstating, never should be same sequence for diferent companies.

yajo commented:
We have double-checked together and indeed the implementation had to change a bit. It had to be a compute instead of a default, because default fields don't have access to current record.

It should be OK now, and it has the migration script. @anddago78 is in charge to add tests before going out of draft.

Functionally speaking, you can review again please. Thanks everyone.

yajo commented on account_journal_general_sequence/models/account_journal.py:
```diff
@@ -13,27 +13,39 @@ class AccountJournal(models.Model):
+                        "company_id": one.company_id.id,
+                        "implementation": "no_gap",
+                        "prefix": "%(range_year)s/",
+                        "padding": 6,
```
Leave it like it was, so you don't have to change the preexisting test. This way you avoid introducing more breaking changes than needed.

```suggestion
                        "padding": 10,
```

yajo commented on account_journal_general_sequence/tests/test_numbering.py:
```diff
@@ -64,10 +64,10 @@ def test_renumber(self):
+        self.assertEqual(opening_invoice.entry_number, "2022/000001")
+        self.assertEqual(old_invoice.entry_number, "2022/000002")
+        self.assertEqual(new_invoice.entry_number, "2022/000003")
+        self.assertEqual(next_year_invoice.entry_number, "2023/000001")
```
Revert these changes too.

yajo commented on account_journal_general_sequence/tests/test_numbering.py:
```diff
@@ -85,3 +85,20 @@ def test_install_no_entry_number(self):
+                "company_id": cmp.id,
+            }
+        )
+        self.env.company_id = cmp.id
```
This pattern can introduce a problem. We are not supposed to alter the company like that. The company is supposed to be obtained from the context, and the context is supposed to be frozen. If you do it like this, it can have unwanted side effects.

Instead, create a new env and assign it:

```suggestion
        self.env = self.env(context=dict(self.env.context, allowed_company_ids=cmp.ids))
```

yajo commented on account_journal_general_sequence/tests/test_numbering.py:
```diff
@@ -85,3 +85,20 @@ def test_install_no_entry_number(self):
+        new_company_invoice = self._create_invoice(
+            date_invoice="2023-07-01", auto_validate=True
+        )
+        self.assertEqual(new_company_invoice.entry_number, "2023/000001")
```
You'll have to change the assertion once you revert the change on the padding number:

```suggestion
        self.assertEqual(new_company_invoice.entry_number, "2023/0000000001")
```

yajo commented on account_journal_general_sequence/tests/test_numbering.py:
```diff
@@ -85,3 +85,20 @@ def test_install_no_entry_number(self):
+
+    @users("admin")
+    def test_new_company_journal(self):
+        cmp = self.env["res.company"].create({"name": "Company100", "currency_id": 2})
```
I think that this test is incomplete.

AFAICS, it is testing that an invoice in a new company's journal starts with number 1. However, to make the test truly work, it should test that _the same thing in another company also starts with 1_. Otherwise, you wouldn't be testing the case of different companies having different default sequences.

Could you please extend a bit the test to cover the full case?

yajo reviewed (changes requested):
A few requests below.

Apart from those, please one extra request: keep the commit message from https://github.com/OCA/account-financial-tools/commit/f530a53e097224e0cc1b3aed51a5f406a7aed413. I can see that your latest commit lost the description, which was helpful.

And also please add me as co-author, as I wrote half of the code here. 😃 See our internal docs on the subject.

rafaelbn commented:
Please @anddago78 rebase! It says: "This branch has conflicts that must be resolved" 😄

rafaelbn commented:
/ocabot rebase

yajo commented:
It seems like the bot is not smart enough to do a merge, and it can only do a rebase. Here you can understand the difference between those two: https://es.stackoverflow.com/a/83625/191469

To solve this:

```sh
git checkout 16.0-fix_account_journal_general_sequence
git pull upstream 16.0 --rebase
```

Git will start putting your new commits on top of those from OCA/16.0 and, as soon as it finds a conflict, it will stop. You will be able to solve the conflict using VSCode. Then you can run `git rebase --continue` and repeat the process until all commits are rebased and all conflicts are solved.

Finally, do `git push -f` once all is OK.

yajo committed:
[FIX] account_journal_general_sequence: fix multicompany

Instead of defining the new sequence as a field default, it is now a compute. This is because the sequence depends on the company, but we don't have the `company_id` field until the record is created.

Reduced the default number of padded zeroes to 8. This is a product decision.

The original implementation didn't make much sense because it allowed the user to set a different sequence per journal, but Odoo already has that kind of sequence. The only purpose of this module is to have a sequence *per company*. To avoid breaking too much, for now, when the journal sequence is the default one, we set it as readonly.

Limit the available sequences in the renumbering wizard. Display only those that you have access by your selected context companies. For some reason, Odoo doesn't filter sequences by company automatically.

@moduon MT-3076

Co-authored-by: Andrea Cattalani <22261939+anddago78@users.noreply.github.com>

rafaelbn commented:
OK, I think now we can merge @yajo , OK?

yajo commented:
Yes, go ahead.

rafaelbn commented:
/ocabot merge patch
